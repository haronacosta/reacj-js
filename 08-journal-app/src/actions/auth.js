import Swal from 'sweetalert2/dist/sweetalert2.js';

import 'sweetalert2/src/sweetalert2.scss';

import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { deleteNotes } from './notes';
import { uiFinishLoading, uiStartLoading } from './ui';

export const startLoginWithEmailAndPassword = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await dispatch(login(user.uid, user.displayName));
        dispatch(uiFinishLoading());
      })
      .catch((err) => {
        dispatch(uiFinishLoading());
        Swal.fire({
          title: 'Error',
          text: err.message,
          icon: 'error'
        });
      });
  };
};

export const startLoginWithGoogle = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        Swal.fire({
          title: 'Error',
          text: err.message,
          icon: 'error'
        });
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: { uid, displayName }
});

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    dispatch(logout());

    dispatch(deleteNotes());
  };
};

export const logout = () => ({
  type: types.logout
});
