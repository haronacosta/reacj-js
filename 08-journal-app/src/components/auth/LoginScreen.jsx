import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startLoginWithEmailAndPassword, startLoginWithGoogle } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

const LoginScreen = () => {
  const dispatch = useDispatch();
  const { isLoading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: '',
    password: ''
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginWithEmailAndPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startLoginWithGoogle());
  };

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form action="post" onSubmit={handleLogin}>
        <input
          className="auth__input"
          type="text"
          placeholder="Your email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          className="auth__input"
          type="password"
          name="password"
          placeholder="Your password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button className="btn btn-primary btn-block" disabled={isLoading} type="submit">
          Login
        </button>
        <hr />

        <div className="auth__social-networks">
          <p>Login with social network</p>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="Google button"
              />
            </div>
            <p className="btn-text">
              <b className="btn">Sign in with google</b>
            </p>
          </div>
        </div>

        <Link className="link" to="/auth/register">
          Create new account
        </Link>
      </form>
    </>
  );
};

export default LoginScreen;
