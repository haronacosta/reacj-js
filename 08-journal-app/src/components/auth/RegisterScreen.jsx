import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeUIError, setUIError } from '../../actions/ui';

import { useForm } from '../../hooks/useForm';
const RegisterScreen = () => {
  const dispatch = useDispatch();

  const { msgError } = useSelector((state) => state.ui);

  const [values, handleInputChange] = useForm({
    userName: '',
    email: '',
    password: '',
    confirmPassword: ''
  });

  const { userName, email, password, confirmPassword } = values;

  const handleRegister = (e) => {
    e.preventDefault();

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(email, password, userName));
    }
  };

  const isFormValid = () => {
    if (userName.trim().length === 0) {
      dispatch(setUIError('User name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setUIError('Email is invalid'));
      return false;
    } else if (password !== confirmPassword || password.length < 6) {
      dispatch(setUIError('Password is invalid'));
      return false;
    }

    dispatch(removeUIError());
    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form action="post" onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}

        <input
          className="auth__input"
          type="text"
          placeholder="Your name"
          name="userName"
          autoComplete="off"
          value={userName}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Your email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          className="auth__input"
          type="password"
          name="password"
          placeholder="Your password"
          value={password}
          onChange={handleInputChange}
        />

        <input
          className="auth__input"
          type="password"
          name="confirmPassword"
          placeholder="Confirm password"
          value={confirmPassword}
          onChange={handleInputChange}
        />

        <button className="btn btn-primary btn-block mb-5" type="submit">
          Create Account
        </button>

        <Link className="link mb-5 mt-5" to="/auth/register">
          Already have an account?
        </Link>
      </form>
    </>
  );
};

export default RegisterScreen;
