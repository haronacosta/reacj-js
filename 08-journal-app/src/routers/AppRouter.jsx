import { BrowserRouter as Router, Switch, Route, Redirect } from 'react-router-dom';
import { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';

import AuthRouter from './AuthRouter';
import JournalScreen from '../components/journal/JournalScreen';
import { firebase } from '../firebase/firebase-config';
import { login } from '../actions/auth';
import PrivateRoute from './PrivateRoute';
import { PublicRoute } from './PublicRoute';
import { startLoadingNotes } from '../actions/notes';

const AppRouter = () => {
  const dispatch = useDispatch();

  const [checking, setChecking] = useState(true);
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged((user) => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName));

        dispatch(startLoadingNotes(user.uid));

        setIsLoggedIn(true);
      } else {
        setIsLoggedIn(false);
      }

      setChecking(false);
    });
  }, [setChecking, dispatch, setIsLoggedIn]);

  if (checking) {
    return <h1>Waiting...</h1>;
  }

  return (
    <div>
      <Router>
        <div>
          <Switch>
            <PublicRoute path={'/auth'} isAuthenticated={isLoggedIn} component={AuthRouter} />

            <PrivateRoute
              path={'/*'}
              exact
              isAuthenticated={isLoggedIn}
              component={JournalScreen}
            />

            <Redirect to="/auth/login" />
          </Switch>
        </div>
      </Router>
    </div>
  );
};

export default AppRouter;
