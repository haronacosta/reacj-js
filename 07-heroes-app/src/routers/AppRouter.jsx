import React from 'react';
import { Routes, Route, BrowserRouter } from 'react-router-dom';

// Components
import LoginScreen from '../components/auth/LoginScreen';
import DashboardRoutes from './DashboardRoutes';
import PrivateRoute from './PrivateRoute';
import PublicRoute from './PublicRoute';

const AppRouter = () => {
  return (
    <BrowserRouter>
      <Routes>
        {/* <Route path='/login' element={<LoginScreen />} /> */}

        <Route
          path='/login'
          element={
            <PublicRoute>
              <LoginScreen />
            </PublicRoute>
          }
        />

        <Route
          path='/*'
          element={
            <PrivateRoute>
              <DashboardRoutes />
            </PrivateRoute>
          }
        />

        {/* <Route path='/*' element={<DashboardRoutes />} /> */}
      </Routes>
    </BrowserRouter>
  );
};

export default AppRouter;
