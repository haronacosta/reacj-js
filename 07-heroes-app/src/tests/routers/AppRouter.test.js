import { mount } from 'enzyme';
import { AuthContext } from '../../auth/authContext';
import AppRouter from '../../routers/AppRouter';

describe('Pruebas en el AppRouter', () => {
  test('Debe de mostrar el componente LoginScreen si el usuario no esta logueado', () => {
    const contextValue = {
      user: {
        logged: false,
      },
    };

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('h1').text().trim()).toBe('Login Screen');
  });

  test('Se debe de mostrar el componente de Marvel si el usuario esta logueado', () => {
    const contextValue = {
      user: {
        logged: true,
        name: 'Haron Acosta',
      },
    };

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();

    // expect(wrapper.find('h1').text().trim()).toBe('Marvel Screen');

    expect(wrapper.find('.navbar').exists()).toBe(true);
  });
});
