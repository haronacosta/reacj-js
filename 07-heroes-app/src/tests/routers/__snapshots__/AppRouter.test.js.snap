// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el AppRouter Debe de mostrar el componente LoginScreen si el usuario no esta logueado 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;

exports[`Pruebas en el AppRouter Se debe de mostrar el componente de Marvel si el usuario esta logueado 1`] = `
Array [
  <nav
    className="navbar navbar-expand-sm navbar-dark bg-dark"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      Asociaciones
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          className="nav-item nav-link"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          className="nav-item nav-link"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          className="nav-item nav-link"
          href="/search"
          onClick={[Function]}
        >
          Search
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <span
          className="nav-link nav-item text-info"
        >
          Haron Acosta
        </span>
        <button
          className="btn nav-item nav-link"
          onClick={[Function]}
        >
          Logout
        </button>
      </ul>
    </div>
  </nav>,
  <div
    className="container"
  >
    <div>
      <h1>
        Marvel Screen
      </h1>
      Array [
        <h2>
          Hero List
        </h2>,
        <p
          className="text-warning"
        >
          Publisher: 
          Marvel Comics
        </p>,
        <div
          className="row row-cols-1 row-cols-md-2 g-4 animate__animated animate__fadeIn"
        >
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Spider Man"
                    className="card-img-top"
                    src="/assets/heroes/marvel-spider.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Spider Man
                  </div>
                  <p
                    className="card-text"
                  >
                    Peter Parker
                  </p>
                  <p
                    className="card-text"
                  >
                    Amazing Fantasy #15
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-spider"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Captain America"
                    className="card-img-top"
                    src="/assets/heroes/marvel-captain.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Captain America
                  </div>
                  <p
                    className="card-text"
                  >
                    Steve Rogers
                  </p>
                  <p
                    className="card-text"
                  >
                    Captain America Comics #1
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-captain"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Iron Man"
                    className="card-img-top"
                    src="/assets/heroes/marvel-iron.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Iron Man
                  </div>
                  <p
                    className="card-text"
                  >
                    Tony Stark
                  </p>
                  <p
                    className="card-text"
                  >
                    Tales of Suspense #39
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-iron"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Thor"
                    className="card-img-top"
                    src="/assets/heroes/marvel-thor.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Thor
                  </div>
                  <p
                    className="card-text"
                  >
                    Thor Odinson
                  </p>
                  <p
                    className="card-text"
                  >
                    Journey into Myster #83
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-thor"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Hulk"
                    className="card-img-top"
                    src="/assets/heroes/marvel-hulk.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Hulk
                  </div>
                  <p
                    className="card-text"
                  >
                    Bruce Banner
                  </p>
                  <p
                    className="card-text"
                  >
                    The Incredible Hulk #1
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-hulk"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Wolverine"
                    className="card-img-top"
                    src="/assets/heroes/marvel-wolverine.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Wolverine
                  </div>
                  <p
                    className="card-text"
                  >
                    James Howlett
                  </p>
                  <p
                    className="card-text"
                  >
                    The Incredible Hulk #180
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-wolverine"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Daredevil"
                    className="card-img-top"
                    src="/assets/heroes/marvel-daredevil.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Daredevil
                  </div>
                  <p
                    className="card-text"
                  >
                    Matthew Michael Murdock
                  </p>
                  <p
                    className="card-text"
                  >
                    Daredevil #1
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-daredevil"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Hawkeye"
                    className="card-img-top"
                    src="/assets/heroes/marvel-hawkeye.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Hawkeye
                  </div>
                  <p
                    className="card-text"
                  >
                    Clinton Francis Barton
                  </p>
                  <p
                    className="card-text"
                  >
                    Tales of Suspense #57
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-hawkeye"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Cyclops"
                    className="card-img-top"
                    src="/assets/heroes/marvel-cyclops.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Cyclops
                  </div>
                  <p
                    className="card-text"
                  >
                    Scott Summers
                  </p>
                  <p
                    className="card-text"
                  >
                    X-Men #1
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-cyclops"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="col animate__animated animate__fadeIn"
          >
            <div
              className="card"
            >
              <div
                className="row g-4"
              >
                <div
                  className="col-4"
                >
                  <img
                    alt="Silver Surfer"
                    className="card-img-top"
                    src="/assets/heroes/marvel-silver.jpg"
                  />
                </div>
                <div
                  className="col-8"
                >
                  <div
                    className="card-title"
                  >
                    Silver Surfer
                  </div>
                  <p
                    className="card-text"
                  >
                    Norrin Radd
                  </p>
                  <p
                    className="card-text"
                  >
                    The Fantastic Four #48
                  </p>
                  <a
                    className="pl-4 btn btn-outline-success"
                    href="/hero/marvel-silver"
                    onClick={[Function]}
                  >
                    Más info
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>,
      ]
    </div>
  </div>,
]
`;
